File: ios/WarpLinkModule.swift
import Foundation
import React
@_spi(ReactNative) import WarpLink

@objc(WarpLinkModule)
public class WarpLinkModule: RCTEventEmitter {

  /// Guards every piece of shared state below.
  ///
  /// The three are written on one thread and read on another, and nothing used
  /// to stand between them. `handleIncomingURL` runs on the main thread, because
  /// a `UIApplicationDelegate` callback is main-thread only and that is the
  /// documented host hook. Every `RCT_EXTERN_METHOD`, `getInitialURL` included,
  /// runs on React Native's module method queue, and `startObserving` and
  /// `stopObserving` are driven from `RCTEventEmitter`'s own exported
  /// `addListener` and `removeListeners`, so they run there too.
  ///
  /// `NSLock` rather than a queue, to match the core SDK, which guards its own
  /// statics with one `NSLock` and the same lock/defer-unlock shape.
  ///
  /// The rule the call sites keep: decide under the lock, act outside it.
  /// `sendEvent` re-enters React Native and must never hold this.
  ///
  /// Bead: warplink-nmyq.
  private static let lock = NSLock()
  private static var pendingURL: String?
  private static var sharedInstance: WarpLinkModule?
  private var hasListeners = false

  override init() {
    super.init()
    WarpLinkModule.lock.lock()
    defer { WarpLinkModule.lock.unlock() }
    WarpLinkModule.sharedInstance = self
  }

  @objc override public static func requiresMainQueueSetup() -> Bool {
    return false
  }

  override public func supportedEvents() -> [String] {
    return ["onWarpLinkDeepLink"]
  }

  override public func startObserving() {
    WarpLinkModule.lock.lock()
    defer { WarpLinkModule.lock.unlock() }
    hasListeners = true
  }

  override public func stopObserving() {
    WarpLinkModule.lock.lock()
    defer { WarpLinkModule.lock.unlock() }
    hasListeners = false
  }

  @objc public static func handleIncomingURL(_ url: URL) {
    // Decide under the lock, emit outside it. Holding the lock across
    // `sendEvent` would hold it across React Native's own dispatch, which is a
    // deadlock waiting for a host that calls back into the module.
    lock.lock()
    let listener = sharedInstance.flatMap { $0.hasListeners ? $0 : nil }
    if listener == nil {
      pendingURL = url.absoluteString
    }
    lock.unlock()

    listener?.sendEvent(
      withName: "onWarpLinkDeepLink",
      body: ["url": url.absoluteString]
    )
  }

  /**
   * The native SDK's own version string, not the npm package's.
   *
   * This bridge pins its native SDKs by hand, and the pin has drifted before,
   * so a host has to be able to ask what it is actually running.
   */
  @objc func getSdkVersion(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(WarpLink.sdkVersion)
  }

  @objc func getInitialURL(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    // One launch carries one link, so the read and the clear are one step.
    // As two steps they were a check-then-act, and the shipped TypeScript has
    // two callers: the automatic cold-start path and the public
    // `WarpLink.getInitialURL()`. Overlapping, both could take the same url.
    resolve(WarpLinkModule.takePendingURL())
  }

  private static func takePendingURL() -> String? {
    lock.lock()
    defer { lock.unlock() }
    let url = pendingURL
    pendingURL = nil
    return url
  }

  @objc func configure(
    _ config: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let apiKey = config["apiKey"] as? String else {
      rejectWithError(
        code: "E_INVALID_API_KEY_FORMAT",
        message: "apiKey is required",
        rejecter: reject
      )
      return
    }

    let defaults = WarpLinkOptions()
    let options = WarpLinkOptions(
      apiEndpoint: (config["apiEndpoint"] as? String) ?? defaults.apiEndpoint,
      debugLogging: (config["debugLogging"] as? Bool) ?? defaults.debugLogging,
      // The React Native layer owns all cold/warm/deferred dispatch, so the core
      // is opted out of its automatic handling to avoid a double-fire (the TS
      // orchestrator calls getInitialURL / checkDeferredDeepLink itself).
      autoDeepLinkHandling: false,
      autoDeferredCheck: false,
      // Host-declared custom link domains, passed through untouched. The core
      // normalizes them and unions them with the Info.plist list and the server
      // list, so JavaScript never has to.
      linkDomains: Self.stringArray(config["linkDomains"]) ?? defaults.linkDomains
    )

    WarpLink.configure(apiKey: apiKey, options: options)
    resolve(nil)
  }

  /**
   * Resolves through the SDK's React Native SPI entry point, not the public
   * `WarpLink.handleDeepLink`.
   *
   * JavaScript (`WarpLink.ts`) owns all dispatch and calls this one native
   * method for every tap it decides to resolve, automatic or manual, so this
   * has to supersede and cancel on every call the way the Android bridge's
   * shared `resolveDeepLink` already does. The public `handleDeepLink` never
   * does that: it is owed its own answer, because a native host calling it
   * directly is not the bridge routing many taps through one method. Using it
   * here left a superseded tap retrying on the wire for its whole budget and
   * able to bill a link JS had already moved past (warplink-3f0f).
   */
  @objc func handleDeepLink(
    _ url: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let linkURL = URL(string: url) else {
      rejectWithError(
        code: "E_INVALID_URL",
        message: "Invalid URL: \(url)",
        rejecter: reject
      )
      return
    }

    WarpLink.resolveFromReactNativeBridge(linkURL) { result in
      switch result {
      case .success(let link):
        resolve(self.serializeDeepLink(link))
      case .failure(let error):
        self.rejectWithWarpLinkError(error, rejecter: reject)
      }
    }
  }

  /**
   * Whether `url` is a WarpLink link the native SDK would claim. Resolves
   * nothing and touches no network.
   *
   * The automatic path in JavaScript asks this BEFORE it stamps its dedupe
   * window or claims a tap, which is what the native SDK does for itself:
   * `WarpLink.open(_:)` returns on `isWarpLinkURL` before
   * `shouldSuppressDuplicate`. Learning a url is foreign afterwards, from the
   * `E_INVALID_URL` that `handleDeepLink` rejects with, arrives after the
   * dedupe state has already changed, which is the defect warplink-0csz
   * reports.
   *
   * Never rejects. A string `URL` cannot parse is not a WarpLink link, and that
   * is `false`, not an error: the caller is asking a yes/no question about a
   * url it did not choose.
   */
  @objc func isWarpLinkUrl(
    _ url: String,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard let linkURL = URL(string: url) else {
      resolve(false)
      return
    }
    resolve(WarpLink.isWarpLinkURL(linkURL))
  }

  @objc func checkDeferredDeepLink(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    WarpLink.checkDeferredDeepLink { result in
      switch result {
      case .success(let deepLink):
        if let link = deepLink {
          resolve(self.serializeDeepLink(link))
        } else {
          resolve(nil)
        }
      case .failure(let error):
        self.rejectWithWarpLinkError(error, rejecter: reject)
      }
    }
  }

  @objc func getAttributionResult(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    guard WarpLink.isConfigured else {
      rejectWithError(
        code: "E_NOT_CONFIGURED",
        message: "SDK not configured",
        rejecter: reject
      )
      return
    }
    if let attribution = WarpLink.attributionResult {
      resolve(serializeDeepLink(attribution))
    } else {
      resolve(nil)
    }
  }

  @objc func isConfigured(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(WarpLink.isConfigured)
  }

  @objc func isAttributionComplete(
    _ resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    resolve(WarpLink.isAttributionComplete)
  }

  // MARK: - Private Helpers

  private func serializeDeepLink(
    _ link: WarpLinkDeepLink
  ) -> NSDictionary {
    return [
      "linkId": link.linkId,
      "destination": link.destination,
      "deepLinkUrl": link.deepLinkUrl as Any,
      // Lower [String: JSONValue] to a Foundation object graph; the JSONValue
      // enum is not Objective-C representable and would otherwise reach JS as
      // opaque boxes (dropping customParams).
      "customParams": link.customParams.mapValues { $0.foundationValue },
      "isDeferred": link.isDeferred,
      "matchType": link.matchType?.rawValue as Any,
      "matchConfidence": link.matchConfidence as Any,
      "matchGuaranteed": link.matchGuaranteed,
    ]
  }

  // Returns nil when the key is absent or is not an array, so the caller can
  // fall back to the SDK default. TypeScript types linkDomains as string[], but
  // an untyped JavaScript caller can still send anything, and a straight
  // `as? [String]` cast drops the WHOLE declaration over one bad entry. Keeping
  // the strings matches what the Android bridge does with the same payload.
  private static func stringArray(_ value: Any?) -> [String]? {
    guard let items = value as? [Any] else { return nil }
    return items.compactMap { $0 as? String }
  }

  private func rejectWithError(
    code: String,
    message: String,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    reject(code, message, nil)
  }

  private func rejectWithWarpLinkError(
    _ error: WarpLinkError,
    rejecter reject: RCTPromiseRejectBlock
  ) {
    reject(
      Self.errorCode(for: error),
      error.errorDescription ?? String(describing: error),
      error as NSError
    )
  }

  private static func errorCode(for error: WarpLinkError) -> String {
    switch error {
    case .notConfigured: return "E_NOT_CONFIGURED"
    case .invalidApiKeyFormat: return "E_INVALID_API_KEY_FORMAT"
    case .invalidApiKey: return "E_INVALID_API_KEY"
    case .networkError: return "E_NETWORK_ERROR"
    case .serverError: return "E_SERVER_ERROR"
    case .invalidURL: return "E_INVALID_URL"
    case .linkNotFound: return "E_LINK_NOT_FOUND"
    case .passwordRequired: return "E_PASSWORD_REQUIRED"
    case .decodingError: return "E_DECODING_ERROR"
    }
  }
}

File: android/src/main/kotlin/app/warplink/reactnative/WarpLinkModule.kt
package app.warplink.reactnative

import android.app.Activity
import android.content.Intent
import android.net.Uri
import com.facebook.react.bridge.ActivityEventListener
import com.facebook.react.bridge.Arguments
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableArray
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.WritableArray
import com.facebook.react.bridge.WritableMap
import com.facebook.react.bridge.WritableNativeArray
import com.facebook.react.bridge.WritableNativeMap
import com.facebook.react.modules.core.DeviceEventManagerModule
import org.json.JSONArray
import org.json.JSONObject
import app.warplink.WarpLink
import app.warplink.WarpLinkDeepLink
import app.warplink.WarpLinkError
import app.warplink.WarpLinkOptions

class WarpLinkModule(
    reactContext: ReactApplicationContext
) : ReactContextBaseJavaModule(reactContext),
    ActivityEventListener {

    /**
     * The cold-start link. See InitialUrlBuffer: it is not a plain field
     * because a host can ask for the link before React Native has attached the
     * Activity, and the answer then has to wait for `onHostResume`.
     */
    private val initialUrl = InitialUrlBuffer(reactContext)

    init {
        reactContext.addActivityEventListener(this)
        initialUrl.capture()
    }

    override fun getName(): String = "WarpLinkModule"

    /**
     * The parameters are NOT nullable, and that is load-bearing.
     *
     * `ActivityEventListener` was Java through React Native 0.79, so Kotlin saw
     * `Intent!` and a nullable override compiled. React Native 0.80 rewrote the
     * interface in Kotlin with non-null parameters, and a nullable override then
     * overrides nothing: the bridge stops compiling entirely, on 0.80 and every
     * release after it. Non-null satisfies the whole supported range, because a
     * Java platform type accepts either.
     *
     * Bead: warplink-y0lt.
     */
    override fun onNewIntent(intent: Intent) {
        // A warm-start link belongs to this event and never to getInitialURL.
        // Without this the same url could be delivered twice: once as the
        // event, once as a late cold-start read of the Activity's now-current
        // intent.
        initialUrl.claimedByEvent()
        intent.data?.let { uri ->
            val params = Arguments.createMap().apply {
                putString("url", uri.toString())
            }
            sendEvent("onWarpLinkDeepLink", params)
        }
    }

    override fun onActivityResult(
        activity: Activity,
        requestCode: Int,
        resultCode: Int,
        data: Intent?
    ) {
        // Not used
    }

    @ReactMethod
    fun getInitialURL(promise: Promise) {
        initialUrl.read(promise)
    }

    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    @ReactMethod
    fun configure(config: ReadableMap, promise: Promise) {
        try {
            val apiKey = config.getString("apiKey")
                ?: return promise.reject(
                    "E_INVALID_API_KEY_FORMAT",
                    "apiKey is required"
                )

            val defaults = WarpLinkOptions()
            val options = WarpLinkOptions(
                apiEndpoint = if (config.hasKey("apiEndpoint"))
                    config.getString("apiEndpoint") ?: defaults.apiEndpoint
                else defaults.apiEndpoint,
                debugLogging = if (config.hasKey("debugLogging"))
                    config.getBoolean("debugLogging")
                else defaults.debugLogging,
                // The React Native layer owns all cold/warm/deferred dispatch, so
                // the core is opted out of its automatic handling to avoid a
                // double-fire with the TS orchestrator.
                automaticDeepLinks = false,
                automaticDeferredDeepLinks = false,
                // Host-declared custom link domains, passed through untouched.
                // The core normalizes them and unions them with the manifest
                // meta-data and the server list, so JavaScript never has to.
                linkDomains = readLinkDomains(config, defaults.linkDomains),
            )

            WarpLink.configure(reactApplicationContext, apiKey, options)
            promise.resolve(null)
        } catch (e: Exception) {
            rejectWithError(e, promise)
        }
    }

    @ReactMethod
    fun handleDeepLink(url: String, promise: Promise) {
        try {
            val uri = Uri.parse(url)
            WarpLink.handleDeepLink(uri) { result ->
                result
                    .onSuccess { deepLink ->
                        if (deepLink != null) {
                            promise.resolve(serializeDeepLink(deepLink))
                        } else {
                            promise.resolve(null)
                        }
                    }
                    .onFailure { error ->
                        rejectWithError(error, promise)
                    }
            }
        } catch (e: Exception) {
            rejectWithError(e, promise)
        }
    }

    /**
     * Whether [url] is a WarpLink link the native SDK would claim. Resolves
     * nothing and touches no network.
     *
     * The automatic path in JavaScript asks this BEFORE it stamps its dedupe
     * window or claims a tap, which is what the native SDK does for itself:
     * `AutoLinkHandler.dispatch` returns on `WarpLink.isWarpLinkUri` before
     * `claimLocked`. Learning a url is foreign afterwards, from the
     * `E_INVALID_URL` that `handleDeepLink` rejects with, arrives after the
     * dedupe state has already changed, which is the defect warplink-0csz
     * reports.
     *
     * Never rejects. A string that is not a url at all is not a WarpLink link,
     * and that is `false`, not an error: the caller is asking a yes/no question
     * about a url it did not choose.
     */
    @ReactMethod
    fun isWarpLinkUrl(url: String, promise: Promise) {
        val answer = try {
            WarpLink.isWarpLinkUri(Uri.parse(url))
        } catch (_: Exception) {
            false
        }
        promise.resolve(answer)
    }

    @ReactMethod
    fun checkDeferredDeepLink(promise: Promise) {
        try {
            WarpLink.checkDeferredDeepLink { result ->
                result
                    .onSuccess { deepLink ->
                        if (deepLink != null) {
                            promise.resolve(serializeDeepLink(deepLink))
                        } else {
                            promise.resolve(null)
                        }
                    }
                    .onFailure { error ->
                        rejectWithError(error, promise)
                    }
            }
        } catch (e: Exception) {
            rejectWithError(e, promise)
        }
    }

    @ReactMethod
    fun getAttributionResult(promise: Promise) {
        if (!WarpLink.isConfigured) {
            return promise.reject("E_NOT_CONFIGURED", "SDK not configured")
        }
        val attribution = WarpLink.attributionResult
        if (attribution != null) {
            promise.resolve(serializeDeepLink(attribution))
        } else {
            promise.resolve(null)
        }
    }

    @ReactMethod
    fun isConfigured(promise: Promise) {
        promise.resolve(WarpLink.isConfigured)
    }

    /**
     * The native SDK's own version string, not the npm package's.
     *
     * This bridge pins `app.warplink:sdk` by hand, and that pin has drifted
     * before, so a host has to be able to ask what it is actually running.
     */
    @ReactMethod
    fun getSdkVersion(promise: Promise) {
        promise.resolve(WarpLink.SDK_VERSION)
    }

    @ReactMethod
    fun isAttributionComplete(promise: Promise) {
        promise.resolve(WarpLink.isAttributionComplete())
    }

    @ReactMethod
    fun addListener(eventName: String) {
        // No-op. NativeEventEmitter lifecycle stub.
    }

    @ReactMethod
    fun removeListeners(count: Int) {
        // No-op. NativeEventEmitter lifecycle stub.
    }

    // TypeScript types linkDomains as string[], but an untyped JavaScript caller
    // can still send anything, and a positional getString() on a non-string entry
    // throws. Read the whole array once and keep the strings: one bad entry drops
    // itself instead of failing the entire configure() call.
    private fun readStringList(array: ReadableArray?): List<String> =
        array?.toArrayList()?.filterIsInstance<String>() ?: emptyList()

    /**
     * `linkDomains` is `string[]` in TypeScript, but a JavaScript caller can hand
     * the bridge a bare string, and `getArray` then throws. The outer catch
     * turned that into E_SERVER_ERROR and left the SDK unconfigured for the
     * whole launch, which misnames the fault and loses every link. One string
     * is one domain. `ReadableMap.getType` is not used because the compile stub
     * in scripts/check-android-bridge.sh does not model it.
     */
    private fun readLinkDomains(config: ReadableMap, fallback: List<String>): List<String> {
        if (!config.hasKey("linkDomains")) return fallback
        return try {
            readStringList(config.getArray("linkDomains"))
        } catch (_: Exception) {
            val single = try { config.getString("linkDomains") } catch (_: Exception) { null }
            if (single.isNullOrBlank()) fallback else listOf(single)
        }
    }

    private fun serializeDeepLink(link: WarpLinkDeepLink): WritableMap {
        val map = WritableNativeMap()
        map.putString("linkId", link.linkId)
        map.putString("destination", link.destination)
        if (link.deepLinkUrl != null) {
            map.putString("deepLinkUrl", link.deepLinkUrl)
        } else {
            map.putNull("deepLinkUrl")
        }
        map.putMap("customParams", toWritableMap(link.customParams))
        map.putBoolean("isDeferred", link.isDeferred)
        if (link.matchType != null) {
            map.putString("matchType", link.matchType!!.name.lowercase())
        } else {
            map.putNull("matchType")
        }
        if (link.matchConfidence != null) {
            map.putDouble("matchConfidence", link.matchConfidence!!)
        } else {
            map.putNull("matchConfidence")
        }
        map.putBoolean("matchGuaranteed", link.matchGuaranteed)
        return map
    }

    private fun toWritableMap(
        params: Map<String, Any?>
    ): WritableMap {
        val map = WritableNativeMap()
        for ((key, value) in params) {
            when (value) {
                is String -> map.putString(key, value)
                is Boolean -> map.putBoolean(key, value)
                is Int -> map.putInt(key, value)
                // Long and other numerics (e.g. JSON numbers that exceed Int)
                // must not be stringified; deliver them as JS numbers.
                is Number -> map.putDouble(key, value.toDouble())
                is JSONObject -> map.putMap(key, jsonObjectToWritableMap(value))
                is JSONArray -> map.putArray(key, jsonArrayToWritableArray(value))
                JSONObject.NULL -> map.putNull(key)
                null -> map.putNull(key)
                else -> map.putString(key, value.toString())
            }
        }
        return map
    }

    private fun jsonObjectToWritableMap(json: JSONObject): WritableMap {
        val map = WritableNativeMap()
        for (key in json.keys()) {
            val value = json.get(key)
            when (value) {
                is String -> map.putString(key, value)
                is Boolean -> map.putBoolean(key, value)
                is Int -> map.putInt(key, value)
                is Number -> map.putDouble(key, value.toDouble())
                is JSONObject -> map.putMap(key, jsonObjectToWritableMap(value))
                is JSONArray -> map.putArray(key, jsonArrayToWritableArray(value))
                JSONObject.NULL -> map.putNull(key)
                else -> map.putString(key, value.toString())
            }
        }
        return map
    }

    private fun jsonArrayToWritableArray(json: JSONArray): WritableArray {
        val array = WritableNativeArray()
        for (i in 0 until json.length()) {
            when (val value = json.get(i)) {
                is String -> array.pushString(value)
                is Boolean -> array.pushBoolean(value)
                is Int -> array.pushInt(value)
                is Number -> array.pushDouble(value.toDouble())
                is JSONObject -> array.pushMap(jsonObjectToWritableMap(value))
                is JSONArray -> array.pushArray(jsonArrayToWritableArray(value))
                JSONObject.NULL -> array.pushNull()
                else -> array.pushString(value.toString())
            }
        }
        return array
    }

    private fun rejectWithError(error: Throwable, promise: Promise) {
        val code = when (error) {
            is WarpLinkError.NotConfigured -> "E_NOT_CONFIGURED"
            is WarpLinkError.InvalidApiKey -> "E_INVALID_API_KEY"
            is WarpLinkError.InvalidApiKeyFormat -> "E_INVALID_API_KEY_FORMAT"
            is WarpLinkError.NetworkError -> "E_NETWORK_ERROR"
            is WarpLinkError.ServerError -> "E_SERVER_ERROR"
            is WarpLinkError.InvalidUrl -> "E_INVALID_URL"
            is WarpLinkError.LinkNotFound -> "E_LINK_NOT_FOUND"
            is WarpLinkError.PasswordRequired -> "E_PASSWORD_REQUIRED"
            is WarpLinkError.DecodingError -> "E_DECODING_ERROR"
            else -> "E_SERVER_ERROR"
        }
        promise.reject(code, error.message, error)
    }
}
